src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs

```diff
@@ -33,55 +31,84 @@ internal abstract partial class AbstractNavigateToSearchService
         private static async Task<ImmutableArray<INavigateToSearchResult>> FindNavigableDeclaredSymbolInfosAsync(
             Project project, Document searchDocument, string pattern, CancellationToken cancellationToken)
         {
-            var containsDots = pattern.IndexOf('.') >= 0;
-            using (var patternMatcher = new PatternMatcher(pattern, allowFuzzyMatching: true))
+            // If the user created a dotted pattern then we'll grab the last part of the name
+            var (patternName, patternContainerOpt) = PatternMatcher.GetNameAndContainer(pattern);
+            var nameMatcher = PatternMatcher.CreatePatternMatcher(patternName, includeMatchedSpans: true, allowFuzzyMatching: true);
+
+            var containerMatcher = patternContainerOpt != null
+                ? PatternMatcher.CreateDotSeperatedContainerMatcher(patternContainerOpt)
+                : null;
+
+            using (nameMatcher)
+            using (containerMatcher)
             {
-                var result = ArrayBuilder<INavigateToSearchResult>.GetInstance();
-                foreach (var document in project.Documents)
+                var nameMatches = ArrayBuilder<PatternMatch>.GetInstance();
+                var containerMatches = ArrayBuilder<PatternMatch>.GetInstance();
+
+                try
                 {
-                    if (searchDocument != null && document != searchDocument)
-                    {
-                        continue;
-                    }
+                    return await FindNavigableDeclaredSymbolInfosAsync(
+                        project, searchDocument, nameMatcher, containerMatcher,
+                        nameMatches, containerMatches, cancellationToken).ConfigureAwait(false);
+                }
+                finally
+                {
+                    nameMatches.Free();
+                    containerMatches.Free();
+                }
+            }
+        }
+
+        private static async Task<ImmutableArray<INavigateToSearchResult>> FindNavigableDeclaredSymbolInfosAsync(
+            Project project, Document searchDocument,
+            PatternMatcher nameMatcher, PatternMatcher containerMatcher,
+            ArrayBuilder<PatternMatch> nameMatches, ArrayBuilder<PatternMatch> containerMatches,
```

Comment:
💭 I don't find it intuitive that these parameters are provided as a "scratch pad" but not actually used to return results to the caller. However, I can't say I haven't used the same pattern myself. Plus the method is `private`.